common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -420,18 +432,88 @@ private void notifyListeners() {
         }
         EventExecutor executor = executor();
         if (executor.inEventLoop()) {
```

Comment:
The transport focused executors (backed by `SingleThreadEventExecutor`) could also get into trouble. For example If there is a promise listener chain:
A (operation completes calls B.success(...))
B (operation completes calls N.success(...))
N (operation completes calls (N+1).success(...))

If N is sufficiently large, or the available stack space is sufficiently small relative to N, we may cause a SOE. We currently have some vulnerability to this, but the stack growth due to promise listener notification is at least capped to `MAX_LISTENER_STACK_DEPTH`.

We may be able to do a similar type of fix in `ImmediateExecutor`, 'EmbeddedEventLoop', 'ImmediateEventExecutor`, transport executors, etc... in Netty ... but there are also other implementations of `Executor` that we have no control over (things similar to guava's `DirectExecutor`). Would you be interested in submitting a PR for this?